cpu/atmega_common/periph/i2c.c

```diff
@@ -0,0 +1,347 @@
+
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* send start condition and slave address */
+    if (_start(address, I2C_FLAG_READ) != 0)
+        return 0;
+
+    for (int i = 0; i < length; i++) {
+        /* Send NACK for last received byte */
+        if ((length-i) == 1)
+            TWCR = (1 << TWEN) | (1 << TWINT);
+        else
+            TWCR = (1 << TWEA) | (1 << TWEN) | (1 << TWINT);
+        DEBUG("Wait for byte %i\n", i+1);
+        /* Wait for TWINT Flag set. This indicates that DATA has been received.*/
+        while (!(TWCR & (1 << TWINT)))
+            {}
+        /* receive data byte */
+        my_data[i] = TWDR;
+        DEBUG("Byte %i received\n", i+1);
+    }
+
+    /* end transmission */
+    _stop();
+
+    return length;
+}
+
+int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, void *data)
+{
+    return i2c_read_regs(dev, address, reg, data, 1);
+}
+
+int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg, void *data, int length)
+{
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0)
+        return 0;
+
+    /* send register address and wait for complete transfer to be finished*/
+    if (_write(&reg, 1) != 1) {
+        _stop();
+        return 0;
+    }
+
+    _stop();
+
+    /* now start a new start condition and receive data */
+    return i2c_read_bytes(dev, address, data, length);
+}
+
+int i2c_write_byte(i2c_t dev, uint8_t address, uint8_t data)
+{
+    return i2c_write_bytes(dev, address, &data, 1);
+}
+
+int i2c_write_bytes(i2c_t dev, uint8_t address, const void *data, int length)
+{
+    int bytes = 0;
+
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0)
+        return 0;
+
+    /* send out data bytes */
+    bytes = _write(data, length);
+
+    /* end transmission */
+    _stop();
+
+    return bytes;
+}
+
+int i2c_write_reg(i2c_t dev, uint8_t address, uint8_t reg, uint8_t data)
+{
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    return i2c_write_regs(dev, address, reg, &data, 1);
+}
+
+int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, const void *data, int length)
+{
+    int bytes = 0;
+
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0)
+        return 0;
+    /* send register address and wait for complete transfer to be finished*/
+    if (_write(&reg, 1))
+        /* write data to register */
+        bytes = _write(data, length);
+    /* finish transfer */
+    _stop();
+    /* return number of bytes send */
+    return bytes;
+}
+
+void i2c_poweron(i2c_t dev)
+{
+    if (dev < I2C_NUMOF) {
+        #if defined (CPU_ATMEGA2560) || defined (CPU_ATMEGA1281)
```

Comment:
similar as for the pull reg config: it would be cleaner to handle the register naming difference in the CPU's configuration